### PR TITLE
Update BridgeTowerForContrastiveLearning

### PR DESCRIPTION
**Pull request by abhiwand**
# What does this PR do?

1. Use return_loss in BridgeTowerForContrastiveLearning
2. Update example in BridgeTowerForContrastiveLearning
3. Handles @amyeroberts suggestion from https://github.com/huggingface/transformers/pull/21964 to use smaller vocab_size for BridgeTowerTester

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@amyeroberts @sgugger can you please help review this fix.

### CODE REVIEW

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 1]
```diff
@@ -1857,15 +1871,22 @@ def forward(
...
 
         if not return_dict:
-            output = tuple(logits)
+            output = (logits, text_embeds, image_embeds, cross_embeds)
+            if output_attentions:
+                output = output + (outputs[4],)
+                if output_hidden_states:
+                    output = output + (outputs[3],)
+            elif output_hidden_states:
+                output = output + (outputs[3],)
```
Comment (ydshieh) [inline src/transformers/models/bridgetower/modeling_bridgetower.py:1888]: I am wondering why `BridgeTowerModelOutput` and `BridgeTowerContrastiveOutput` will have `attentions` and `hidden_states` swapped, and causing the somehow strange order here.

IMO, it's much better to have a consistent ordering. I would also like  hear @amyeroberts 's opinion.

Comment (ydshieh) [reply src/transformers/models/bridgetower/modeling_bridgetower.py:1888]: Also, I can't comment at the exact line , but the block below  is better to put the arguments in the order that is defined in `BridgeTowerContrastiveOutput`. In terms of functionality, it is fine, but putting in the same order will reduce some (mental) burden.

```python
        return BridgeTowerContrastiveOutput(
            attentions=outputs.attentions,
            hidden_states=outputs.hidden_states,
            text_embeds=text_embeds,
            image_embeds=image_embeds,
            cross_embeds=cross_embeds,
            logits=logits,
            loss=itc_loss,
        )
```

**Review by ydshieh — COMMENTED** [review]
Hi @abhiwand Thank you for working on this.

Here is the 1st part of my reviews. I will give the reviews for the tests part.

**Review by ydshieh — COMMENTED** [review]
Regarding tests

File: src/transformers/models/bridgetower/modeling_bridgetower.py (hunk 1 above)
Comment (amyeroberts) [reply src/transformers/models/bridgetower/modeling_bridgetower.py:1888]: Yes, I agree, it's better to have consistent ordering. I think most, if not all, other model outputs return hidden_states then attentions

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 2]
```diff
@@ -166,6 +166,8 @@ class BridgeTowerContrastiveOutput(ModelOutput):
     Output type of ['BridgeTowerForContrastiveLearning']
 
     Args:
+        loss (`torch.FloatTensor` of shape `(1,)`, *optional*, returned when `labels` is provided):
```
[the remark this answers is not in the document]
Comment (tileintel) [reply src/transformers/models/bridgetower/modeling_bridgetower.py]: We addressed this in the latest commit. Thanks for pointing this out. @ydshieh

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 3]
```diff
@@ -1789,12 +1789,11 @@ def forward(
         output_attentions: Optional[bool] = None,
         output_hidden_states: Optional[bool] = True,
         return_dict: Optional[bool] = None,
-        labels: Optional[torch.LongTensor] = None,
+        return_loss: Optional[bool] = True,
```
[the remark this answers is not in the document]
Comment (tileintel) [reply src/transformers/models/bridgetower/modeling_bridgetower.py]: We addressed this in the latest commit. Thanks for suggestion.

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 4]
```diff
@@ -1789,12 +1789,11 @@ def forward(
...
     ) -> Union[BridgeTowerContrastiveOutput, Tuple[torch.FloatTensor]]:
         r"""
-        labels (`torch.LongTensor` of shape `(batch_size, 1)`, *optional*):
-            Labels for computing the image-text matching loss. 0 means the pairs don't match and 1 means they match.
-            The pairs with 0 will be skipped for calculation.
+        return_loss (`bool`, *optional*):
+            Whether or not to return the contrastive loss. Default is True.
```
[the remark this answers is not in the document]
Comment (tileintel) [reply src/transformers/models/bridgetower/modeling_bridgetower.py]: We fixed this in the latest commit. Thanks for suggestion.

[the remark this answers is not in the document]
Comment (tileintel) [reply src/transformers/models/bridgetower/modeling_bridgetower.py]: We fixed this in the latest commit. Thanks for suggestion.

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 5]
```diff
@@ -1803,14 +1802,29 @@ def forward(
...
+        >>> outputs = model(**inputs)
+
+        >>> inputs = processor(images, texts[::-1], padding=True, return_tensors="pt")
+        >>> outputs_swapped = model(**inputs)
+
+        >>> print("Loss", outputs.loss.item())
+        Loss 0.00191505195107311
+
+        >>> print("Loss with swapped images", outputs_swapped.loss.item())
+        Loss with swapped images 2.1259872913360596
```
[the remark this answers is not in the document]
Comment (tileintel) [reply src/transformers/models/bridgetower/modeling_bridgetower.py]: We addressed this in the latest commit. Thanks for suggestion.

File: src/transformers/models/bridgetower/modeling_bridgetower.py (hunk 1 above)
Comment (tileintel) [reply src/transformers/models/bridgetower/modeling_bridgetower.py]: We fixed this in the latest commit. Thanks for suggestion.

**Comment by tileintel** [comment]
Thank @ydshieh and @amyeroberts for your suggestions. In the latest commit, we addressed all of your feedbacks except one that regards to tests inspired from Clip's tests.
@ydshieh Regarding tests, yes, you are right, it will require more works. We plan to have another PR to improve tests (following your suggestion) soon, yet this is not the main purpose of this PR. We would love to have you help us working on this, we truly appreciate your offer to help.

File: tests/models/bridgetower/test_modeling_bridgetower.py [hunk 6]
```diff
@@ -135,6 +137,7 @@ def get_config(self):
             output_hidden_states=self.output_hidden_states,
             contrastive_hidden_size=self.contrastive_hidden_size,
             logit_scale_init_value=self.logit_scale_init_value,
+            text_config=text_config,
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply tests/models/bridgetower/test_modeling_bridgetower.py:140]: Let's revert the change regarding `text_config = {"vocab_size": self.vocab_size}` and `text_config=text_config,` here, and address the test in another PR.

`self.vocab_size = 99` is OK to be kept.

**Review by ydshieh — COMMENTED** [review]
Thank you @tileintel . Just 2 places to update, and we are good to merge 🚀

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for adding these updates! 

Once the changes @ydshieh suggested have been included I think we're good to merge :)

**Comment by tileintel** [comment]
Thank @amyeroberts for approving this PR.
@ydshieh We have updated the 2 positions that you most recently suggested to change in the latest commit. Could you please review, approve, and merge the PR? 
We are looking forward to having this PR merged soon. 
Thanks a lot.

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 7]
```diff
@@ -1857,23 +1871,23 @@ def forward(
...
-            hidden_states=outputs.hidden_states,
+            loss=itc_loss,
+            logits=logits,
             text_embeds=text_embeds,
             image_embeds=image_embeds,
             cross_embeds=cross_embeds,
-            logits=logits,
-            loss=itc_loss,
+            attentions=outputs.attentions,
+            hidden_states=outputs.hidden_states,
```
Comment (ydshieh) [inline src/transformers/models/bridgetower/modeling_bridgetower.py:1892]:
```suggestion
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,            
```

Comment (ydshieh) [reply src/transformers/models/bridgetower/modeling_bridgetower.py:1892]: The tests (currently disabled) all pass (launched manually). Good to merge once the above change is committed.

**Review by ydshieh — APPROVED** [review]
@tileintel Could you accept the commit suggestion. Then we are good to go!

File: src/transformers/models/bridgetower/modeling_bridgetower.py (hunk 7 above)
Comment (ydshieh) [reply src/transformers/models/bridgetower/modeling_bridgetower.py:1892]: Let me commit directly @tileintel as it's minor. Hope you are OK with it.

**Comment by ydshieh** [comment]
Thank you again, @tileintel  and @abhiwand, for the work! Merge now 🚀 !

File: src/transformers/models/bridgetower/modeling_bridgetower.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tileintel) [reply src/transformers/models/bridgetower/modeling_bridgetower.py:1891]: Thank @ydshieh for helping this.

**Comment by ydshieh** [comment]
> and please feel free to contact/tag me if you would like me to cooperate on improving/restructuring tests. Thanks

Hi @tileintel. If you are willing and have time to work on the tests part, it would be really great. But as I mentioned earlier, we understand this part is not always the main interest of community contributors: as long as the basic necessary tests are added, it is sufficient for a model addition PR. So I can work on the test restructuring unless you would like to do it :-). WDYT?

**Comment by abhiwand** [comment]
> > and please feel free to contact/tag me if you would like me to cooperate on improving/restructuring tests. Thanks
> 
> Hi @tileintel. If you are willing and have time to work on the tests part, it would be really great. But as I mentioned earlier, we understand this part is not always the main interest of community contributors: as long as the basic necessary tests are added, it is sufficient for a model addition PR. So I can work on the test restructuring unless you would like to do it :-). WDYT?

@ydshieh thanks for pointing this out. We would really appreciate if you can work on restructuring tests. Thank you!

**Comment by ydshieh** [comment]
> > > and please feel free to contact/tag me if you would like me to cooperate on improving/restructuring tests. Thanks
> > 
> > 
> > Hi @tileintel. If you are willing and have time to work on the tests part, it would be really great. But as I mentioned earlier, we understand this part is not always the main interest of community contributors: as long as the basic necessary tests are added, it is sufficient for a model addition PR. So I can work on the test restructuring unless you would like to do it :-). WDYT?
> 
> @ydshieh thanks for pointing this out. We would really appreciate if you can work on restructuring tests. Thank you!

Sure, thanks for letting me know 😊

**Pull request merged**